encoding.bs

```diff
@@ -94,35 +90,133 @@ bitwise AND by "&amp;", and bitwise OR by "|".
 
 <hr>
 
-<p>A <dfn id=concept-token>token</dfn> is a piece of data, such as a <a>byte</a> or
-<a>scalar value</a>.
+<p>An <dfn id=concept-stream export>I/O queue</dfn> is a type of <a for=/>list</a>
+with <a for=list>items</a> of a particular type (i.e. <a>bytes</a> or <a>scalar
+values</a>). <dfn id="end-of-stream" export>End-of-queue</dfn> is a special
+<a for=list>item</a> that can be present in <a for=/>I/O queues</a> of any type, and that
+signifies no more <a for=list>items</a> are in the queue.
+
+<p>To <dfn for="I/O queue" export>create</dfn> an <a for=/>I/O queue</a>, optionally initialized
+with a list <var>contents</var>, return a new <a for=/>I/O queue</a> containing the
+<a for=list>items</a> of <var>contents</var> in order, if it is given, followed by an
+<a>end-of-queue</a> item.
+
+<div class=note>
+ <p>There are two ways to use an <a for=/>I/O queue</a>: in immediate mode, to
+ represent I/O data stored in memory, and in streaming mode, to represent data
+ coming in from the network. Immediate queues have <a>end-of-queue</a> as their
+ last item, whereas streaming queues need not have it, and so their
+ <a for="I/O queue">read</a> operation might block.
+
+ <p>An immediate <a for=/>I/O queue</a> is to be created with the <a for="I/O queue">create</a> an
+ <a for=/>I/O queue</a> operation. A streaming <a for=/>I/O queue</a> is to be created
+ empty, and it is expected that new <a for=list>items</a> will be
+ <a for="I/O queue">pushed</a> to it as data comes in from the network. When the
+ underlying network stream closes, an <a>end-of-queue</a> item is to be
+ <a for="I/O queue">pushed</a> into the queue.
+
+ <p>Since reading from a streaming <a for=/>I/O queue</a> might block, streaming
+ <a for=/>I/O queues</a> are not to be used from an <a for=/>event loop</a>.
+ They are to be used <a>in parallel</a> instead.
+</div>
+
+<p>To <dfn id=concept-stream-read for="I/O queue" export>read</dfn> an
+<a for=list>item</a> from an <a for=/>I/O queue</a> <var>ioQueue</var>, run these steps:
+
+<ol>
+ <li><p>If <var>ioQueue</var> is <a for=list>empty</a>, then wait until its
+ <a for=list>size</a> is at least 1.
+
+ <li><p>If the first item in <var>ioQueue</var> is <a>end-of-queue</a>, then
+ return <a>end-of-queue</a>.
+
+ <li><p>Otherwise, <a for=list>remove</a> the first item in <var>ioQueue</var>
+ and return it.
+</ol>
+
+<p>To read a number <var>number</var> of <a for=list>items</a> from <var>ioQueue</var>,
```

Comment:
This needs a dfn if we are to keep it.

Reply:
Should I use something like `<dfn lt="read" for="I/O queue">read an item</dfn>` vs `<dfn for="I/O queue">read a number of items</dfn>`?

Reply:
Thinking about this some more maybe overloading is the way to go. In that case we should xref "read" here and leave it at that I think. (I tried thinking of some alternative ways of doing this including "read"/"read multiple" (looks awkward) or "read one or more" (but then return type is annoying), and I can't really figure it out.

Perhaps @domenic has thoughts on this.